File: problems-extend-adapt/3901-4000/3902_alternating-level-sums/solution.go
func alternatingLevelSums(root *TreeNode) []int64 {
	frontier := []*TreeNode{root}
	answer := []int64{}
	odd := true
	for len(frontier) > 0 {
		var total int64
		for step := 0; step < len(frontier); step++ {
			index := step
			if !odd {
				index = len(frontier) - 1 - step
			}
			node := frontier[index]
			required := node.Left
			if !odd {
				required = node.Right
			}
			if required == nil {
				break
			}
			total += int64(node.Val)
		}
		answer = append(answer, total)
		next := make([]*TreeNode, 0, 2*len(frontier))
		for _, node := range frontier {
			if node.Left != nil {
				next = append(next, node.Left)
			}
			if node.Right != nil {
				next = append(next, node.Right)
			}
		}
		frontier = next
		odd = !odd
	}
	return answer
}

File: problems-extend-adapt/3901-4000/3902_alternating-level-sums/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public long[] alternatingLevelSums(TreeNode root) {
        List<TreeNode> frontier = new ArrayList<>();
        frontier.add(root);
        List<Long> sums = new ArrayList<>();
        boolean odd = true;
        while (!frontier.isEmpty()) {
            long total = 0;
            for (int step = 0; step < frontier.size(); ++step) {
                int index = odd ? step : frontier.size() - 1 - step;
                TreeNode node = frontier.get(index);
                TreeNode required = odd ? node.left : node.right;
                if (required == null) break;
                total += node.val;
            }
            sums.add(total);
            List<TreeNode> next = new ArrayList<>();
            for (TreeNode node : frontier) {
                if (node.left != null) next.add(node.left);
                if (node.right != null) next.add(node.right);
            }
            frontier = next;
            odd = !odd;
        }
        long[] answer = new long[sums.size()];
        for (int i = 0; i < answer.length; ++i) answer[i] = sums.get(i);
        return answer;
    }
}
